src/exec.cpp

```diff
@@ -325,13 +325,29 @@ static void run_internal_process_or_short_circuit(parser_t &parser, const std::s
     }
 }
 
+bool blocked_signals_for_job(const job_t &job, sigset_t *sigmask) {
+    // Block some signals in background jobs for which job control is turned off (#6828).
+    if (!job.is_foreground() && !job.wants_job_control()) {
+        sigaddset(sigmask, SIGINT);
+        sigaddset(sigmask, SIGQUIT);
+        return true;
+    }
+    return false;
+}
+
 /// Call fork() as part of executing a process \p p in a job \j. Execute \p child_action in the
 /// context of the child. Returns true if fork succeeded, false if fork failed.
 static bool fork_child_for_process(const std::shared_ptr<job_t> &job, process_t *p,
                                    const dup2_list_t &dup2s, const char *fork_type,
                                    const std::function<void()> &child_action) {
     pid_t pid = execute_fork();
     if (pid == 0) {
+        sigset_t sigmask;
```

Comment:
Let's unify signal manipulation under `child_setup_process` - see the call to `signal_reset_handlers`. I think you can probably just replace that call outright with one that sets a sigmask, which you compute.